Ton Gateway (#588)
zeta-chain/docs: src/pages/developers/chains/ton.mdx

To interact with universal applications from TON, use the TON gateway.

TON gateway supports:

- Depositing TON to a universal app or an account on ZetaChain
- Depositing TON and calling a universal app
- Withdrawing TON from ZetaChain

## Deposit TON

To deposit TON to an EOA or a universal contract, send an internal message to
the Gateway contract with the following structure:

> ```fc

Reviewer: Let's use one-liner (TON's notation)
```suggestion
```func
op_code:uint32 query_id:uint64 evm_recipient:slice (160 bits)
```

;; Message body structure
;; op_code: 32 bits
;; query_id: 64 bits
;; evm_recipient: 160 bits (20 bytes)
```

> The `op_code` should be set to `101` for deposits.

Reviewer: ```suggestion
The deposit `op_code` is `101`. `query_id` is reserved for future use, leave it to `0`.
```

Reviewer: We can follow up with an example payload in TS:
```ts
const opDeposit = 101
const body = beginCell()
    .storeUint(opDeposit, 32)
    .storeUint(0, 64)
    .storeUint(zevmRecipient, 160)
    .endCell();
```

Author: 0709f96

The `evm_recipient` specifies the address on ZetaChain that will receive the
deposited TON. This can be either an externally-owned account (EOA) or a
universal app address.

After the deposit is processed, the receiver receives the [ZRC-20
version](/developers/tokens/zrc20) of the deposited TON.

## Deposit TON and Call a Universal App

To deposit TON and call a universal app contract, send an internal message to
the Gateway contract with the following structure:

```fc

> ;; Message body structure

Reviewer: ```suggestion
op_code:uint32 query_id:uint64 evm_recipient:slice (160 bits) call_data:cell
```

;; op_code: 32 bits
;; query_id: 64 bits
;; evm_recipient: 160 bits (20 bytes)
;; call_data: cell (up to 2KB)
```

> The `op_code` should be set to `102` for deposit and call operations.

Reviewer: ```suggestion
The depositAndCall `op_code` is `102`. `query_id` is reserved for future use, leave it to `0`. Also note that call_data should be a cell encoded in ["snake data"](https://docs.ton.org/v3/guidelines/dapps/asset-processing/nft-processing/metadata-parsing#snake-data-encoding) format (supported by most TON libraries)
```

Reviewer: Example:
```ts
// @ton/core. Up to 2 kilobytes
const callData = stringToCell("lorem ipsum ......");

const opDepositAndCall = 102;

const body = beginCell()
            .storeUint(opDepositAndCall, 32)
            .storeUint(0, 64)
            .storeUint(zevmRecipient, 160)
            .storeRef(callData)
            .endCell();

The `evm_recipient` must be the address of a universal app contract.

The `call_data` cell contains the payload that will be passed to the `onCall`
function of the universal app contract.

After the cross-chain transaction is processed, the `onCall` function of the
universal app contract is executed.

## Gas Fees

The Gateway contract charges gas fees for different operations:

- Deposit: 10,000 gas units
- Deposit and Call: 13,000 gas units

> - Authority operations: 20,000 gas units

Reviewer: We can omit this

Author: e1d0279

To interact with universal applications from TON, use the TON gateway.

TON gateway supports:

- Depositing TON to a universal app or an account on ZetaChain
- Depositing TON and calling a universal app

> - Withdrawing TON from ZetaChain

Author: I'll address this in a follow up PR and add this to other chains (Solana & Sui) as well.
